[package] nlohmann_json/*: includedirs not set in layout

### Description

The recipe does not set the `self.cpp.source.includedirs` property inside the `layout` function and `basic_layout` also does not set it.

### Package and Environment Details

* Package Name/Version: **nlohmann_json/3.11.3**
* Operating System+version: **Debian 12 (bookworm)**
* Compiler+version: **Clang 14**
* Docker image: **debian/12**
* Conan version: **conan 2.19.0**
* Python version: **Python 3.11.2**


### Conan profile

Host profile:
[settings]
arch=x86_64
build_type=Release
compiler=clang
compiler.libcxx=libstdc++11
compiler.version=14
os=Linux
[conf]
tools.build:compiler_executables={'c': 'clang-14', 'cpp': 'clang++-14'}
tools.build:jobs=16
tools.cmake.cmake_layout:build_folder_vars=['settings.arch', 'settings.os']
tools.cmake.cmaketoolchain:generator=Ninja
tools.system.package_manager:mode=install
tools.system.package_manager:sudo=True
tools.system.package_manager:tool=apt-get

Build profile:
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=12
os=Linux
[conf]
tools.build:jobs=16
tools.cmake.cmake_layout:build_folder_vars=['settings.arch', 'settings.os']
tools.cmake.cmaketoolchain:generator=Ninja
tools.system.package_manager:mode=install
tools.system.package_manager:sudo=True
tools.system.package_manager:tool=apt-get

### Steps to reproduce

```
conan editable add recipes/nlohmann_json/all/ --version 3.11.3
conan source recipes/nlohmann_json/all/ --version 3.11.3
conan test recipes/nlohmann_json/all/test_package nlohmann_json/3.11.3
```

### Logs

<details><summary>Click to expand log</summary>

```
CMake Error in CMakeLists.txt:
  Imported target "nlohmann_json::nlohmann_json" includes non-existent path

    "/workspace/conan-build-recipe/recipes/nlohmann_json/all/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
```

</details>

--- a/conan/tools/layout/__init__.py
+++ b/conan/tools/layout/__init__.py
@@ -14,3 +14,4 @@ def basic_layout(conanfile, src_folder=".", build_folder=None):
     conanfile.folders.generators = os.path.join(conanfile.folders.build, "conan")
     conanfile.cpp.build.bindirs = ["."]
     conanfile.cpp.build.libdirs = ["."]
+    conanfile.cpp.source.includedirs = ["include"]

--- a/conan/tools/layout/__init__.py
+++ b/conan/tools/layout/__init__.py
@@ -14,4 +14,6 @@ def basic_layout(conanfile, src_folder=".", build_folder=None):
     conanfile.folders.generators = os.path.join(conanfile.folders.build, "conan")
     conanfile.cpp.build.bindirs = ["."]
     conanfile.cpp.build.libdirs = ["."]
-    conanfile.cpp.source.includedirs = ["include"]
+
+    if not conanfile.cpp.source.includedirs:
+        conanfile.cpp.source.includedirs = ["include"]

--- a/test/integration/toolchains/gnu/test_basic_layout.py
+++ b/test/integration/toolchains/gnu/test_basic_layout.py
@@ -29,7 +29,7 @@ def test_basic_layout_subproject(basic_layout, expected_path):
                                        "conanautotoolstoolchain.{}".format(ext)))
 
 
-def test_editable_includedirs():
+def test_editable_includes():
     c = TestClient()
     dep = textwrap.dedent("""
         from conan import ConanFile
@@ -54,3 +54,31 @@ def test_editable_includedirs():
     dep_path = os.path.join(c.current_folder, "dep", "src", "include")
     assert f"SYSTEM_HEADER_SEARCH_PATHS_dep_dep[config=Release][arch=x86_64][sdk=*] = \"{dep_path}\"" in xcode
 
+
+def test_editable_includes_previously_defined():
+    # if someone already defined self.cpp.source.includedirs make sure we don't overwrite
+    # their value
+    c = TestClient()
+    dep = textwrap.dedent("""
+        from conan import ConanFile
+        from conan.tools.layout import basic_layout
+
+        class Dep(ConanFile):
+            name = "dep"
+            version = "0.1"
+            def layout(self):
+                self.cpp.source.includedirs = ["somefolder"]
+                basic_layout(self, src_folder="src")
+            """)
+    c.save({"dep/conanfile.py": dep,
+            "pkg/conanfile.py": GenConanfile("pkg", "0.1").with_settings("build_type", "arch")
+                                                          .with_requires("dep/0.1")})
+    c.run("editable add dep")
+    c.run("install pkg -s arch=x86_64 -g CMakeDeps -g PkgConfigDeps -g XcodeDeps")
+    data = c.load(f"pkg/dep-release-x86_64-data.cmake")
+    assert 'set(dep_INCLUDE_DIRS_RELEASE "${dep_PACKAGE_FOLDER_RELEASE}/src/somefolder")' in data
+    pc = c.load("pkg/dep.pc")
+    assert "includedir=${prefix}/src/somefolder" in pc
+    xcode = c.load("pkg/conan_dep_dep_release_x86_64.xcconfig")
+    dep_path = os.path.join(c.current_folder, "dep", "src", "somefolder")
+    assert f"SYSTEM_HEADER_SEARCH_PATHS_dep_dep[config=Release][arch=x86_64][sdk=*] = \"{dep_path}\"" in xcode

--- a/test/integration/toolchains/gnu/test_basic_layout.py
+++ b/test/integration/toolchains/gnu/test_basic_layout.py
@@ -3,6 +3,7 @@ import platform
 import textwrap
 import pytest
 
+from conan.test.assets.genconanfile import GenConanfile
 from conan.test.utils.tools import TestClient
 
 @pytest.mark.parametrize("basic_layout, expected_path", [
@@ -26,3 +27,30 @@ def test_basic_layout_subproject(basic_layout, expected_path):
     ext = "sh" if platform.system() != "Windows" else "bat"
     assert os.path.isfile(os.path.join(c.current_folder, "pkg", expected_path, "conan",
                                        "conanautotoolstoolchain.{}".format(ext)))
+
+
+def test_editable_includedirs():
+    c = TestClient()
+    dep = textwrap.dedent("""
+        from conan import ConanFile
+        from conan.tools.layout import basic_layout
+
+        class Dep(ConanFile):
+            name = "dep"
+            version = "0.1"
+            def layout(self):
+                basic_layout(self, src_folder="src")
+            """)
+    c.save({"dep/conanfile.py": dep,
+            "pkg/conanfile.py": GenConanfile("pkg", "0.1").with_settings("build_type", "arch")
+                                                          .with_requires("dep/0.1")})
+    c.run("editable add dep")
+    c.run("install pkg -s arch=x86_64 -g CMakeDeps -g PkgConfigDeps -g XcodeDeps")
+    data = c.load(f"pkg/dep-release-x86_64-data.cmake")
+    assert 'set(dep_INCLUDE_DIRS_RELEASE "${dep_PACKAGE_FOLDER_RELEASE}/src/include")' in data
+    pc = c.load("pkg/dep.pc")
+    assert "includedir=${prefix}/src/include" in pc
+    xcode = c.load("pkg/conan_dep_dep_release_x86_64.xcconfig")
+    dep_path = os.path.join(c.current_folder, "dep", "src", "include")
+    assert f"SYSTEM_HEADER_SEARCH_PATHS_dep_dep[config=Release][arch=x86_64][sdk=*] = \"{dep_path}\"" in xcode
+

